crates/crates_io_docs_rs/src/lib.rs

```diff
@@ -0,0 +1,214 @@
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use test_case::test_case;
+
+    async fn mock(
+        krate: &str,
+        version: &str,
+        status: StatusCode,
+    ) -> (mockito::ServerGuard, mockito::Mock) {
+        let mut server = mockito::Server::new_async().await;
+        let mock = server
+            .mock("POST", &format!("/crate/{krate}/{version}/rebuild")[..])
+            .match_header("Authorization", "Bearer test_token")
+            .with_status(StatusCode::CREATED.as_u16().into())
+            .with_status(status.as_u16().into());
+
+        (server, mock)
+    }
+
+    #[tokio::test]
+    async fn test_ok() -> anyhow::Result<()> {
+        let (server, mock) = mock("krate", "0.1.0", StatusCode::CREATED).await;
+        mock.create();
+
+        let docs_rs = RealDocsRsClient::new(server.url(), "test_token")?;
+
+        docs_rs.rebuild_docs("krate", "0.1.0").await?;
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_crate_not_found() -> anyhow::Result<()> {
+        let (server, mock) = mock("krate", "0.1.0", StatusCode::NOT_FOUND).await;
+        mock.create();
+
+        let docs_rs = RealDocsRsClient::new(server.url(), "test_token")?;
+
+        assert!(matches!(
+            docs_rs.rebuild_docs("krate", "0.1.0").await,
+            Err(DocsRsError::NotFound)
+        ));
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_crate_too_many_requests() -> anyhow::Result<()> {
+        let (server, mock) = mock("krate", "0.1.0", StatusCode::TOO_MANY_REQUESTS).await;
+        mock.create();
+
+        let docs_rs = RealDocsRsClient::new(server.url(), "test_token")?;
+
+        assert!(matches!(
+            docs_rs.rebuild_docs("krate", "0.1.0").await,
+            Err(DocsRsError::RateLimited)
+        ));
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    #[test_case(StatusCode::UNAUTHORIZED)]
+    #[test_case(StatusCode::FORBIDDEN)]
+    async fn test_permissions(status: StatusCode) -> anyhow::Result<()> {
+        let (server, mock) = mock("krate", "0.1.0", status).await;
+        mock.create();
+
+        let docs_rs = RealDocsRsClient::new(server.url(), "test_token")?;
+
+        assert!(matches!(
+            docs_rs.rebuild_docs("krate", "0.1.0").await,
+            Err(DocsRsError::Unauthorized)
+        ));
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_bad_request_message() -> anyhow::Result<()> {
+        let (server, mock) = mock("krate", "0.1.0", StatusCode::BAD_REQUEST).await;
+        let body = serde_json::to_vec(&serde_json::json!({
+            "message": "some error message"
+        }))?;
+        mock.with_body(&body).create();
+
+        let docs_rs = RealDocsRsClient::new(server.url(), "test_token")?;
+
+        if let Err(DocsRsError::BadRequest(msg)) = docs_rs.rebuild_docs("krate", "0.1.0").await {
+            assert_eq!(msg, "some error message");
+        } else {
+            panic!("Expected BadRequest error");
+        }
```

Comment:
we're using https://crates.io/crates/claims in a couple of our packages, which should simplify this assertion a little bit :)